### PR TITLE
*: a little fix and add 3 functions.

### PR DESCRIPTION
**Pull request by hicqu**
1. add function `TableInfo.ColumnInIndex` in model;
2. add function `Schema.FindColumnAndIndex` in expression;
3. add function `rewriteWithPreprocess` in plan.

EXPLAIN:
1) is a new function.
2) is for simplify those cases: first call `FindColumn`, and then call `ColumnIndex`, such as `buildUpdateLists` in **plan/logical_plan_builder.go**.
3) is for this case: we need rewrite `func_call(a, b, c)` to `func_call(db1.table1.a, db2.table2.b,  db3.table3.c)`, but we can't (or don't want) modify that ast tree inplace. For example, if the ast tree is a global large object, and it's unsafe to modify that inplace.

### CODE REVIEW

**Comment by hicqu** [comment]
@shenli @tiancaiamao @XuHuaiyu , PTAL, thanks!

File: expression/schema.go [hunk 1]
```diff
@@ -74,6 +74,13 @@ func (s *Schema) Clone() *Schema {
 // FindColumn finds an Column from schema for a ast.ColumnName. It compares the db/table/column names.
 // If there are more than one result, it will raise ambiguous error.
 func (s *Schema) FindColumn(astCol *ast.ColumnName) (*Column, error) {
+	col, _, err := s.FindColumnAndIndex(astCol)
+	return col, err
```
Comment (shenli) [inline expression/schema.go]: errors.Trace(err)

File: model/model.go [hunk 2]
```diff
@@ -135,6 +135,18 @@ func (t *TableInfo) GetPkName() CIStr {
 	return CIStr{}
 }
 
+// ColumnIsInIndex tests c is included in any indices of t or not.
```
Comment (shenli) [inline model/model.go]: // ColumnIsInIndex checks whether c is included in any indices of t.

File: model/model.go [hunk 3]
```diff
@@ -135,6 +135,18 @@ func (t *TableInfo) GetPkName() CIStr {
...
+func (t *TableInfo) ColumnIsInIndex(c *ColumnInfo) bool {
+	for _, index := range t.Indices {
+		for _, column := range index.Columns {
+			if column.Name == c.Name {
```
Comment (shenli) [inline model/model.go]: Consider c.Name.L?

File: plan/expression_rewriter.go [hunk 4]
```diff
@@ -61,12 +61,20 @@ func evalAstExpr(expr ast.ExprNode, ctx context.Context) (types.Datum, error) {
 // And this function returns a result expression, a new plan that may have apply or semi-join.
 func (b *planBuilder) rewrite(expr ast.ExprNode, p LogicalPlan, aggMapper map[*ast.AggregateFuncExpr]int, asScalar bool) (
 	expression.Expression, LogicalPlan, error) {
+	return b.rewriteWithPreprocess(expr, p, aggMapper, asScalar, nil)
+}
+
+// rewriteWithPreprocess can be used for if we need to adjust expr but can't do that inplace.
```
Comment (XuHuaiyu) [inline plan/expression_rewriter.go]: 1. for handling the situation that if
2. change `,` to `.`

Comment (XuHuaiyu) [inline plan/expression_rewriter.go]: I don't really understand what does `adjust expr` mean here?

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**